### PR TITLE
Fix(pages): pages dev crash leaving port open

### LINKED ISSUE #2445: 🐛 BUG: `wrangler pages dev` crashes on Worker compile error, leaving port occupied
### What version of `Wrangler` are you using?

2.6.2

### What operating system are you using?

Mac

### Describe the Bug

When running the `wrangler pages dev` command, if the Worker fails compilation (due to coding error or partial save), will make the whole process fail, making you restart the command.

The bigger problem is that upon restart `0.0.0.0:8788` is marked as occupied for a variable amount of time, making the restart impossible, unless the port is changed.

### LINKED ISSUE #3378: 🐛 BUG: Pages Function Compilation Error Crashes Servers, Leaves Ports Listening
### Which Cloudflare product(s) does this pertain to?

Pages, Wrangler

### What version of `Wrangler` are you using?

3.0.1

### What operating system are you using?

Linux -- Fedora 37 & 38

### Describe the Bug

Exact replica of #94, which is strange as that issue was patched 1.5 years ago.

When using Wrangler (local development, so Miniflare included) and editing a Page function, the server crashes on compilation errors instead of showing an error message as @GregBrimble configured. The server's listener on port 8788 is left running and has to be manually killed before restarting the server. This behavior is not apparently in Wrangler when developer a Worker.

Reproduction Steps -- and [minimal reproducible example](https://github.com/arimgibson/pages-functions-repro):
1. Use `pnpm create cloudflare` to create a new Pages application (used Astro, framework shouldn't matter)
2. Add a functions directory with `helloworld.js`, which has working onRequest function
3. Start server with `pnpm pages:dev`, visit http://127.0.0.0:8788/helloworld
4. Change helloworld.js to throw compliation error, such as removing an ending parenthesis

Screenshot from 2023-05-30 14-27-57

Screenshot from 2023-05-30 14-29-14

### LINKED ISSUE #3619: 🐛 BUG: `wrangler pages dev` crushes when `_worker.js/index.js` is invalid
### Which Cloudflare product(s) does this pertain to?

Pages

### What version(s) of the tool(s) are you using?

wrangler 3.2.0

### What version of Node are you using?

16

### What operating system are you using?

Mac

### Describe the Bug

If I run `wrangler pages dev` and have an `_worker.js` directory with a `index.js` file, if the file is invalid this causes wrangler to crush (it also still runs the dev server under the hood without releasing the port so I cannot even simply just restart the process).

___

This is different from what happens if `_worker.js` is a file, in that case it `wrangler pages dev` does handle the situation correctly and doesn't crush (it temporarily errors, but the dev server is still working on the previous iteration and after the file is fixed all is back to normal).

___

I think that this is a relevant bug to fix because a watch build process writing into the `_worker.js/index.js` file will make it temporary invalid after each change, completely breaking the dx for the user (which would need to manually start/kill `wrangler pages dev` and/or kill the underlying unattached dev server and/or re-running wrangler on different ports constantly).

### Please provide a link to a minimal reproduction

https://github.com/dario-piotrowicz/wrangler-pages-dev-workerjs-dir-bug-reproduction

### LINKED ISSUE #3663: 🐛 BUG: Wrangler crashes when running `pages dev` if functions contain invalid syntax
### Which Cloudflare product(s) does this pertain to?

Pages, Wrangler core

### What version(s) of the tool(s) are you using?

Wrangler 3.3.0

### What version of Node are you using?

18.17.0

### What operating system are you using?

WSL (Ubuntu 22.04)

### Describe the Bug

When running `wrangler pages dev [...]`, Typescript functions get auto-transpiled as part of the dev watcher process.

If the TS file contains invalid syntax, the build errors as expected.

When this happens, however, the dev server itself appears to crash and require a restart. I would not expect this. Furthermore, the dev server doesn't shut down cleanly from this crash, instead leaving the old HTTP server live on `:8788` (and proxy/proxy port if appropriate), requiring the process to be manually killed to restart the dev server.

### Please provide a link to a minimal reproduction

_No response_

### Please provide any relevant error logs

image
image

### LINKED ISSUE #4031: 🐛 Bug Report — Pages Functions - workerd crashes on syntax errors
refs https://github.com/cloudflare/workers-sdk/issues/3663

When working with pages functions in dev mode, ie:

```sh
wrangler pages dev public
```

When you edit a file and get a syntax error, wrangler crashes. And worse than just crashing, it leaves the workerd process running so you can't just start it up again. You have to find the process ID, then `kill` the process before you can start it up again.

The error can be as simple as having an extra bracket, eg:

image

### PR DESCRIPTION
**Pull request by JacksonKearl**
Fixes #3663, fixes #3378, fixes #2445, fixes #3619, fixes #4031

**What this PR solves / how to test:**
Previously, `wrangler pages dev` would watch for file updates and crash any time the functions directory had a syntax error. Now, it prints an error message describing the error and continues to serve the old version of the function until the error is fixed. 

Test by running the sample repo [here](https://github.com/JacksonKearl/cf-minimal-sample/tree/main):
- `npx wrangler pages dev -- npx vite dev`, then add a syntax error to `functions/api.ts` (a random `}` works well). 
- Observe that the process crashes and the ports opened by `vite` and `wrangler` are both left open. 
- Now, do the same but working from this branch. Observe that the process does not crash, an error message is logged to the console describing the issue, and the server continues to serve the old version of functions. 
- Now fix the syntax error, observe the server live-reloads to the new function version. 

Before:
```
➜  sample git:(main) wrangler pages dev -- npx vite dev 
Running npx vite dev...
Sleeping 5 seconds to allow proxy process to start before attempting to automatically determine port...
To skip, specify the proxy port with --proxy.
[proxy]: 
  VITE v4.4.8  ready in 98 ms


[proxy]:   ➜  Local:   http://localhost:5173/
  ➜  Network: use --host to expose

Automatically determined the proxy port to be 5173.
▲ [WARNING] No compatibility_date was specified. Using today's date: 2023-08-02.


Compiling worker to "/var/folders/lx/nq6ssgds30x3rgq2m88gxznh0000gn/T/functionsWorker-0.9335350627688517.mjs"...
✨ Compiled Worker successfully
 ⛅️ wrangler 3.4.0
------------------
wrangler dev now uses local mode by default, powered by 🔥 Miniflare and 👷 workerd.
To run an edge preview session for your Worker, use wrangler dev --remote
⎔ Starting local server...
[mf:wrn] The latest compatibility date supported by the installed Cloudflare Workers Runtime is "2023-07-24",
but you've requested "2023-08-02". Falling back to "2023-07-24"...
[mf:inf] Ready on http://127.0.0.1:8788/
```
Now add the syntax error:
```
✘ [ERROR] Unexpected "}"

    functions/api.ts:5:0:
      5 │ }
        ╵ ^

/Users/jacksonkearl/.npm-global/lib/node_modules/wrangler/wrangler-dist/cli.js:30632
            throw a;
            ^

Error: Build failed with 1 error:
functions/api.ts:5:0: ERROR: Unexpected "}"
    at failureErrorWithLog (/Users/jacksonkearl/.npm-global/lib/node_modules/wrangler/node_modules/esbuild/lib/main.js:1591:15)
    at /Users/jacksonkearl/.npm-global/lib/node_modules/wrangler/node_modules/esbuild/lib/main.js:1047:28
    at /Users/jacksonkearl/.npm-global/lib/node_modules/wrangler/node_modules/esbuild/lib/main.js:992:67
    at buildResponseToResult (/Users/jacksonkearl/.npm-global/lib/node_modules/wrangler/node_modules/esbuild/lib/main.js:1045:7)
    at /Users/jacksonkearl/.npm-global/lib/node_modules/wrangler/node_modules/esbuild/lib/main.js:1157:14
    at responseCallbacks.<computed> (/Users/jacksonkearl/.npm-global/lib/node_modules/wrangler/node_modules/esbuild/lib/main.js:694:9)
    at handleIncomingPacket (/Users/jacksonkearl/.npm-global/lib/node_modules/wrangler/node_modules/esbuild/lib/main.js:749:9)
    at Socket.readFromStdout (/Users/jacksonkearl/.npm-global/lib/node_modules/wrangler/node_modules/esbuild/lib/main.js:670:7)
    at Socket.emit (node:events:513:28)
    at addChunk (node:internal/streams/readable:324:12) {
  errors: [
    {
      detail: undefined,
      id: '',
      location: {
        column: 0,
        file: 'functions/api.ts',
        length: 1,
        line: 5,
        lineText: '}',
        namespace: '',
        suggestion: ''
      },
      notes: [],
      pluginName: '',
      text: 'Unexpected "}"'
    }
  ],
  warnings: []
}

```
Now fix the syntax error and try to re-run:
```
➜  sample git:(main) ✗ wrangler pages dev -- npx vite dev
Running npx vite dev...
Sleeping 5 seconds to allow proxy process to start before attempting to automatically determine port...
To skip, specify the proxy port with --proxy.
[proxy]: Port 5173 is in use, trying another one...

[proxy]: 
  VITE v4.4.8  ready in 102 ms


[proxy]:   ➜  Local:   http://localhost:5174/
  ➜  Network: use --host to expose

Automatically determined the proxy port to be 5174.
▲ [WARNING] No compatibility_date was specified. Using today's date: 2023-08-02.


Compiling worker to "/var/folders/lx/nq6ssgds30x3rgq2m88gxznh0000gn/T/functionsWorker-0.8054426095938616.mjs"...
✨ Compiled Worker successfully
 ⛅️ wrangler 3.4.0
------------------
wrangler dev now uses local mode by default, powered by 🔥 Miniflare and 👷 workerd.
To run an edge preview session for your Worker, use wrangler dev --remote
⎔ Starting local server...
[mf:wrn] The latest compatibility date supported by the installed Cloudflare Workers Runtime is "2023-07-24",
but you've requested "2023-08-02". Falling back to "2023-07-24"...
workerd/util/symbolizer.c++:99: warning: Not symbolizing stack traces because $LLVM_SYMBOLIZER is not set. To symbolize stack traces, set $LLVM_SYMBOLIZER to the location of the llvm-symbolizer binary. When running tests under bazel, use `--test_env=LLVM_SYMBOLIZER=<path>`.
*** Fatal uncaught kj::Exception: kj/async-io-unix.c++:945: failed: ::bind(sockfd, &addr.generic, addrlen): Address already in use; toString() = 0.0.0.0:8788
stack: 105bbfebf 105bbfc57 105bbd4ef 10433f60b 10433fed4 10430891f 10430c32b 10430c15f 10430c147 105bf5ddb 105bf6037 105bf478f 105bf455f 1043001ab 18e20be4f
✘ [ERROR] MiniflareCoreError [ERR_RUNTIME_FAILURE]: The Workers runtime failed to start. There is likely additional logging output above.
```

After:
```
[mf:inf] Ready on http://127.0.0.1:8788/
[proxy]: 
  VITE v4.4.8  ready in 97 ms


[proxy]:   ➜  Local:   http://localhost:5173/
  ➜  Network: use --host to expose
```
Now add the syntax error:
```
✘ [ERROR] Unexpected "}"

    functions/api.ts:5:0:
▲ [WARNING] Unexpected error building Functions directory: ./functions - Build failed with 1 error:

  functions/api.ts:5:0: ERROR: Unexpected "}"

```
Now fix the syntax error and observe the automatic reload:
```
✨ Compiled Worker successfully
⎔ Reloading local server...
[mf:inf] Updated and ready on http://127.0.0.1:8788/
```

**Associated docs issue(s)/PR(s):**

**Author has included the following, where applicable:**

- [ ] Tests
- [x] Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))

**Reviewer is to perform the following, as applicable:**

- Checked for inclusion of relevant tests
- Checked for inclusion of a relevant changeset
- Checked for creation of associated docs updates
- Manually pulled down the changes and spot-tested

**Note for PR author:**

We want to celebrate and highlight awesome PR review! If you think this PR received a particularly high-caliber review, please assign it the label `highlight pr review` so future reviewers can take inspiration and learn from it.

### CODE REVIEW

**Comment by JacksonKearl** [comment]
@Skye-31 sure, I have added it

**Comment by RamIdeas** [comment]
cc @GregBrimble can we add an e2e test to cover this? To specifically ensure we recover from syntax errors

**Comment by jadbox** [comment]
I hit this same issue today with WSL and it took me nearly four hours to track it down to this issue. Wrangler needs to report if the socket is already open!

**Comment by JacksonKearl** [comment]
@jadbox In my experience wrangler simply crashes when the port is taken. What did you observe?

**Comment by jadbox** [comment]
@JacksonKearl I get no message either, even with verbose level logging. It's why it took me nearly 4 hours of troubleshoot... 🫠

**Comment by treeder** [comment]
Any estimate on when this will be released?

**Comment by JacobMGEvans** [comment]
Be sure to run `pnpm checks` locally, and `pnpm prettify`, there are some linter errors that need resolving. Great PR btw thanks for your contribution! 
Screenshot 2023-09-26 at 8 50 39 AM

File: packages/wrangler/e2e/dev.test.ts [hunk 1]
```diff
@@ -117,6 +117,57 @@ describe("basic dev tests", () => {
...
+
+			// Output during error state is indeterminate.
+			// const { text: text2 } = await retry(
+			// 	(s) => s.status !== 200 || s.text === "Hello World!",
+			// 	async () => {
+			// 		const r = await fetch(`http://127.0.0.1:${port}`);
+			// 		return { text: await r.text(), status: r.status };
+			// 	}
+			// );
+			// expect(text2).toMatchInlineSnapshot('"Hello World!"');
```
Comment (RamIdeas) [inline packages/wrangler/e2e/dev.test.ts:149]: For this test to be worth running, we need to confirm the syntax error took effect. Without these steps, the seed function is called immediately after the last and so could lead to a race condition as to whether wrangler/miniflare/workerd even saw the syntax error

Can you describe what you mean by "indeterminate"? Maybe just asserting the response is _not_ "Hello World" is enough?

File: packages/wrangler/src/pages/dev.ts [hunk 2]
```diff
@@ -376,6 +390,11 @@ export const Handler = async ({
 						logger.warn(
 							getFunctionsNoRoutesWarning(functionsDirectory, "skipping")
 						);
+					}
+					if (e instanceof FunctionsBuildError) {
```
Comment (Skye-31) [inline packages/wrangler/src/pages/dev.ts:394]:
```suggestion
					} else if (e instanceof FunctionsBuildError) {
```

**Comment by JacksonKearl** [comment]
@RamIdeas I don't think the test is worth running. I only added it because I've had this PR that fixes 5 separate issues that are encountered *constantly* for anyone doing development in a wrangler project up for 2 months with absolutely 0 traction besides to say "this doesn't have a test".

**Review by penalosa — APPROVED** [review]

**Review by RamIdeas — APPROVED** [review]

**Review by WalshyDev — APPROVED** [review]

**Pull request merged**